rust/src/dataset/fragment.rs

```diff
@@ -76,6 +78,8 @@ impl FileFragment {
         let filename = format!("{}.lance", Uuid::new_v4());
         let fragment = Fragment::with_file(id as u64, &filename, &schema);
 
+        progress.begin(&fragment).await?;
```

Comment:
Is the value the user is getting is that they get to know the data file name? Just trying to understand why a user couldn't just wrap calls to `Fragment::create()` with these hooks themselves if they want this.

Reply:
`Fragment::create()` is a block call, when it returns, the data file is finished. This is used to capture the half-written data.

An alternative would be to have a `Fragment::open_writer() -> FragmentWriter`, and let user to do

```
let writer = Fragment::open(path)?
for batch in chunked_stream:
     writer.write(batch)
writer.close()
```

Reply:
Okay so IIUC, the reason you can't just have the user write:

```rust
before();
let fragment = Fragment::create(...).await?;
complete(&fragment);
```

Is that if this crashes during `Fragment::create()`, it could create a half-written file that we want to cleanup, but without the special hook they won't know the filename?

BTW, to cleanup for S3, you would need to know the multipart upload ID to call [Abort Multipart Upload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html). object_store returns this as part of `put_multipart`. Knowing the object path won't help, since the object isn't created until [CompleteMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) is called. Perhaps we could also pass the multipart upload id, if there is one?

But this could work for local filesystems.

Reply:
> Is that if this crashes during Fragment::create(), it could create a half-written file that we want to cleanup, but without the special hook they won't know the filename?

Yes, that's true. 

>BTW, to cleanup for S3, you would need to know the multipart upload ID to call [Abort Multipart Upload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html). object_store returns this as part of put_multipart. Knowing the object path won't help, since the object isn't created until [CompleteMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) is called. Perhaps we could also pass the multipart upload id, if there is one?

Does it expose via `object_store` crate? Should we add upstream support for exposing such info? Also, what would be the case for Google cloud storage?

Reply:
[put_multipart](https://docs.rs/object_store/0.7.0/object_store/trait.ObjectStore.html#tymethod.put_multipart) returns the upload id. There's even an [abort_multipart](https://docs.rs/object_store/0.7.0/object_store/trait.ObjectStore.html#tymethod.abort_multipart) function you can pass it to. Same thing for GCS. For Azure, there's no need since they automatically clean up after 7 days.

Though, if this is all just for clean up, I wonder why we can just have the user call vacuum if there were any failed tasks? Saves them the trouble of keeping track of these files themselves.

Reply:
In an environment without coordinators, i.e. using `SLURM`, you might want to find out what fragments have failed to write, and re-submit the job to re-run these fragments again.

Maybe there are other alternatives to find out which fragments to dirty, and needs to re-run?

Reply:
> In an environment without coordinators, i.e. using SLURM, you might want to find out what fragments have failed to write, and re-submit the job to re-run these fragments again.

It makes sense that distributed applications want this. What I'm trying to figure out is why this should be part of Lance itself, and can't be part of the users code instead. Exposing details like which data files are being written, or their multi-part ids, could get messy later. 

From what I can tell, all you need is each worker to send a message when they start, then send a message when they have successfully finished. (The messages could be sent to some queue, or just be written as objects to S3, or something else.) Then the user can look for which workers send the start message and not the finished message, and rerun those.

Reply:
> What I'm trying to figure out is why this should be part of Lance itself, and can't be part of the users code instead. 

Which is fine. So it needs the `Fragment::writer()` to expose which data file is being written.

```
before();
let fragment = Fragment::create(...).await?;  <--- if we crash here, no tracking of dirty files / multi-uploadID
complete(&fragment);
```